src/Psalm/Codebase.php

```diff
@@ -1967,9 +1967,7 @@ public function addTaintSource(
 
         $this->taint_flow_graph->addSource($source);
 
-        $expr_type->parent_nodes = [
-            $source->id => $source,
-        ];
+        $expr_type->parent_nodes[$source->id] = $source;
```

Comment:
This isn't really that closely related to everything else in this PR, but I'm pretty sure it's a correct change. I'm not familiar enough with the taint feature to know how to test it, but I worry that without this change some of the taint stuff might mess up the `parent_nodes` for a reference.

Reply:
Let's fix that when we have a reported bug then. I see we have multiple instances of both ways of declaring parent_nodes and I'm not sure when to choose which

Reply:
I was leaning that way originally, but it looks like this only exists for plugins to use, and based on the name it's supposed to **add** a source, not remove all other sources as well.
I think the only time `parent_nodes` should be set to an array rather than setting an offset is when the previous nodes are known to be wrong, such as when assigning a variable.

```
$foo = 'a';
$foo = 'b'; // Should no longer have a parent node for `$foo = 'a'`
```

Reply:
I'm unclear on what can cause a variable to have multiple parents though

Reply:
I'm a bit unsure about that myself. I utilized multiple parents for references, but outside of that I don't know what use it has, other than it probably has something to do with taints.